oximeter/db/src/oxql/ast/table_ops/filter.rs

```diff
@@ -152,82 +335,269 @@ impl Filter {
     // This is the maximum timestamp, before which any filtered point must lie.
     // This is used to determine the query end time.
     pub(crate) fn last_timestamp(&self) -> Option<DateTime<Utc>> {
-        match self {
-            Filter::Atom(atom) => atom.last_timestamp(),
-            Filter::Expr(expr) => expr.last_timestamp(),
+        match &self.expr {
+            FilterExpr::Simple(inner) => inner.last_timestamp(),
+            FilterExpr::Compound(inner) => inner.last_timestamp(),
         }
     }
 
     // Return the name of all identifiers listed in this filter.
     fn ident_names(&self) -> BTreeSet<&str> {
-        match self {
-            Filter::Atom(atom) => {
+        match &self.expr {
+            FilterExpr::Simple(inner) => {
                 let mut out = BTreeSet::new();
-                out.insert(atom.ident.as_str());
+                out.insert(inner.ident.as_str());
                 out
             }
-            Filter::Expr(expr) => {
-                let mut all = expr.left.ident_names();
-                all.extend(expr.right.ident_names());
+            FilterExpr::Compound(inner) => {
+                let mut all = inner.left.ident_names();
+                all.extend(inner.right.ident_names());
                 all
             }
         }
     }
+
+    fn is_xor(&self) -> bool {
+        self.is_op(LogicalOp::Xor)
+    }
+
+    fn is_op(&self, expected_op: LogicalOp) -> bool {
+        let FilterExpr::Compound(CompoundFilter { op, .. }) = &self.expr else {
+            return false;
+        };
+        op == &expected_op
+    }
+
+    // If this is an XOR, rewrite it to a disjunction of conjunctions.
+    //
+    // If it is not, return a clone of self.
+    fn rewrite_xor_to_disjunction(&self) -> Self {
```

Comment:
We can express it in SQL, but it's not easy to tell which parts of the query are independent in that case. The goal of the DNF craziness is to have a bunch of disjunctions, each of which can be run separately to get the group of timeseries keys consistent with it. So we rewrite `a ^ b` into `(a && !b) || (!a && b)` so that we have two such groups.

I could also be missing something though, and this isn't needed!

Reply:
I understand the need to transform a query in a distributed conjunctive form into a distributed disjunctive form. But if there is no distribution in play because the operation can be executed as a primitive in the target execution environment (SQL in this case) do we really need a transformation that factors a disjunctive operation?

Reply:
I hope I'm not misunderstanding your question, but I think we do need to rewrite XORs.

The purpose of the conversion to DNF is not because ClickHouse has any limitations in its SQL engine. All of the logical expressions OxQL supports can be written directly in ClickHouse SQL. We're doing the transformation so that the OxQL layer can figure out how to merge the data from the field tables with those from the measurements table, **without** first doing a complete join between them. I tried that with the SQL prototype, and it falls over pretty quickly as the data gets larger.

The example in one of my long-winded comments above is relevant here. Suppose you want to select all of one timeseries, and part of another. (That's not an XOR, but it is an OR, so has similar properties.) To do that, you have two choices:

- Run a complete inner join between the field tables and the measurements table, and then just select whatever was written in the OxQL query. No transformation needed! This is how the SQL prototype from a few months works, and it's great, until the data gets to even moderate size. At that point, the JOIN is **enormous** -- it takes up a huge amount of memory to duplicate all the fields for every measurement, and then we have to pay the cost to move all those over the network.
- Somehow select the subset you care about from both the field tables and measurement table, and only transfer those.

That latter choice is what I'm going for here. But to do that, you need to know which records from the field table the query implies, _and_ which time ranges those records correspond to. Those time ranges can be different for different subsets of the field table, as they are if you want to select all of one timeseries and part of another. The code needs to store a mapping from (list of keys -> time range).

The disjunctive normal form is about know how to define each list of keys. Every disjunction is independent, so it is logically correct to independently fetch the keys for each of those, and map that to the time range that corresponds to it.

It's true that an XOR is a primitive in the target environment. It's also logically equivalent to an OR, meaning there _could_ be two independent time ranges at play here. So I'm rewriting the XOR to an OR so that we can make sure we don't accidentally merge the wrong data.

I think there's a good optimization opportunity here! The conversion to DNF is really only needed if there is more than one time range at play in the query. If not, then we can do the original version of the query without any conversion. I certainly think we should do that at a later time, but it is an optimization and I wanted to get it correct first.

I hope that answers the question!

Reply:
Ok, I think I'm starting to slowly come around. Maybe the thing that caused me to look at this a bit sideways was the fact that we're unconditionally factoring out the xor. But I _think_ I see the issue where the xor is a part of a more complex expression and we need to fully normalize everything to ensure we get the filtering correct. Sorry for the confusion, still trying to wrap my head around the semantics.

Reply:
Not at all, it's subtle and took me a long time to be able to even describe the problem! I'll draw up a figure too, a visual explanation might be more helpful.